src/hopen/renderer/xf.cljc

```diff
@@ -136,6 +136,19 @@
 (defn- inline-quote [env val]
   val)
 
+(defn- inline-or [env & args]
+  (reduce #(if (tpl-eval env %2)
+             (reduced true)
+             false)
+          args))
+
+(defn- inline-and [env & args]
+  (reduce #(if (tpl-eval env %2)
+             true
+             (reduced false))
+          args))
+
```

Comment:
I'm not sure what you mean by unrelated branch. `and` and `or` were added in the course in implementing the parser. They are referenced [here](https://github.com/clojure-tw/hopen/blob/605af718ba9522301856e23c018e873e3dcb2924/src/hopen/syntax/mustache.cljc#L359), [here](https://github.com/clojure-tw/hopen/blob/605af718ba9522301856e23c018e873e3dcb2924/src/hopen/syntax/mustache.cljc#L361), and [here](https://github.com/clojure-tw/hopen/blob/605af718ba9522301856e23c018e873e3dcb2924/src/hopen/syntax/mustache.cljc#L375).

Reply:
I used the wrote wording. They are related, but adding new functions to the default environment is not the subject of your PR.

PR are either accepted or rejected (and maybe updated and reviewed again). There is no in-between like taking some of the parts of the commits. For this reason, I advise opening a separate PR for modifying stuffs which are not the main purpose of the main PR. It would be easier to discuss, possibly amend and been merged.

The idea is similar to the single responsibility objects or function: it is better not to have more than 1 reason for the object / function to be obsolete and de refactored - for the PR it means to only do 1 subject at a time. The more subjects in a PR, the higher the probability of having 1 subject on which people would disagree, preventing the merge for any of the code addressing any of the subjects.

Reply:
For this parser, having `and` and `or` is a required part of making it actually work so it is well within the "subject" of this branch.

This parser is the first concrete parser to actually make use of and fully exercise the hopen renderer. It's only natural to find that there are many missing pieces in the default environment that are useful or required to support the parsers. 

Following what you're describing, it would take time and effort a to create a new PR, fully explain why it's needed, wait for code review, and approval, just so it can eventually be merged back into *this* branch so the work can continue. Why would I do that? How much additional time do you want me to put into stuff like this?

I'm the only one working on this particular parser, in my free-time, for the (hopefully) fun of being able to work with some people in clojure locally. I don't understand your insistence that I jump through multiple hoops to satisfy some kind of strange PR structure that isn't explained anywhere and is outside of common github practices.

Reply:
> For this parser, having `and` and `or` is a required part of making it actually work so it is well within the "subject" of this branch.

I understand that.

> This parser is the first concrete parser to actually make use of and fully exercise the hopen renderer. It's only natural to find that there are many missing pieces in the default environment that are useful or required to support the parsers.

I agree that.

> Following what you're describing, it would take time and effort a to create a new PR, fully explain why it's needed, wait for code review, and approval, just so it can eventually be merged back into _this_ branch so the work can continue. Why would I do that? How much additional time do you want me to put into stuff like this?

That's why I was suggesting that you let me know, so I can add it faster.

When I add things in the renderer's default environment, I am considering all the parsers at the same time, and need to carefully choose what should be in the defaults and should not. That's why I prefer it to be done independently from the parsers, in a separate branch.

> I'm the only one working on this particular parser, in my free-time, for the (hopefully) fun of being able to work with some people in clojure locally. I don't understand your insistence that I jump through multiple hoops to satisfy some kind of strange PR structure that isn't explained anywhere and is outside of common github practices.

If it not fun anymore, I suggest you to stop working on it. I know that my standard for quality is way higher than the average and it can be very frustrating for you to have some code that starts to work but that don't get merged in the project yet.

I would have been more willing to merge this PR if it was contained in its own namespace and not affecting the rest of the code. At the moment, this PR:
- Added the source code of backtick in its root.
- Added a test framework that modifies the test output's format.
- Added some code to the renderer.
- Added some variable to the default environment.

A PR being merged or not merge (with nothing in between), I don't want to merge it as it is now.

.. and I am not sure if addressing all those listed issues would solve the problem. There are other problems with the source code, I think it should be simpler.

In the Gitter chat room, I was proposing you to use the "Segment" type from the instaparse library which let you have substrings which keep references to the original string with an index and size information. Using it would make a big chunk of you source code simpler.

I did not fully understood how you use the line numbers in your parser, but if it is just for error reporting, you may as well calculate them once you need to output an error, it may result in a more decoupled source code.

My last feedback would be more an advice, it's to give instaparse a try. It has the potential to make a big part of your code simpler.

Reply:
It’s been a strange short journey for sure. The 4 points that you cited as the reason for not accepting the PR are all quite superficial and extremely minor problems.

It really is no big deal if you don’t accept the PR. I wouldn’t mind in the least even if we eventually decide to throw out this implementation because things could be done much more easily and cleanly in instaparse. Its the learning, the discussion, and the journey that makes it fun.

What I do mind, is that although the original invitation was to work on something “together” so the community has something to discuss over meetups, you’ve now assigned yourself the boss. Your expect your opinions to be unquestioningly followed for some reason.

Since you want to make this whole process intentionally difficult and this has definitely ceased being “fun”, I’m closing the PR.